internal/twirptest/service_test.go

```diff
@@ -1527,3 +1527,123 @@ func TestPanicFlushing(t *testing.T) {
 		t.Errorf("twirp client err has unexpected message %q, want %q", twerr.Msg(), "Internal service panic")
 	}
 }
+
+type errReader int
+
+func (errReader) Read(p []byte) (n int, err error) {
+	return 0, fmt.Errorf("test error")
+}
+
+func TestContextCancelError(t *testing.T) {
+	// test case validate context cancel error handling
+	var s Haberdasher
+	server := NewHaberdasherServer(s, nil)
+
+	ctx := context.Background()
+	ctx, cancel := context.WithCancel(ctx)
+
+	// Make a request, that will endpoint method
+	req, _ := http.NewRequest(http.MethodPost, "http://testing:8080/twirp/twirp.internal.twirptest.Haberdasher/MakeHat", errReader(0))
+	req.Header.Set("Accept", "application/protobuf")
+	req.Header.Set("Content-Type", "application/protobuf")
+	// Associate the cancellable context we just created to the request
+	req = req.WithContext(ctx)
+	// cancel context
+	cancel()
+	w := httptest.NewRecorder()
+	server.ServeHTTP(w, req)
+
+	// validate error code
+	expectedErrCode := twirp.ServerHTTPStatusFromErrorCode(twirp.Canceled)
+	if w.Code != expectedErrCode {
+		t.Errorf("twirp ErrorCode expected to be %q, but found %q", expectedErrCode, w.Code)
+	}
+
+	// validate error message
+	respBytes, err := ioutil.ReadAll(w.Body)
+	if err != nil {
+		t.Fatalf("Could not even read bytes from response: %q", err.Error())
+	}
+
+	expectedErrMessage := "context canceled"
+	if !strings.Contains(string(respBytes), expectedErrMessage) {
+		t.Errorf("twirp client err has unexpected message %q, want %q", string(respBytes), expectedErrMessage)
+	}
+}
+
+func TestDeadlineExceededError(t *testing.T) {
+	// test case validate context deadline exceed error handling
+	var s Haberdasher
+	server := NewHaberdasherServer(s, nil)
+
+	ctx := context.Background()
+	ctx, _ = context.WithTimeout(ctx, 1*time.Millisecond)
+
+	// Make a request, that will endpoint method
+	req, _ := http.NewRequest(http.MethodPost, "http://testing:8080/twirp/twirp.internal.twirptest.Haberdasher/MakeHat", errReader(0))
+	req.Header.Set("Accept", "application/protobuf")
+	req.Header.Set("Content-Type", "application/protobuf")
+	// Associate the cancellable context we just created to the request
+	req = req.WithContext(ctx)
+
+	// wait for deadline exceed timeout in context object
+	select {
+	case <-ctx.Done():
+		// ctx.Err() will have context.DeadlineExceeded
+	case <-time.After(1 * time.Second):
+		t.Fatalf("Waited too long for DeadlineExceeded ")
+	}
+	w := httptest.NewRecorder()
```

Comment:
It difficult to generate error while reading body `buf, err := ioutil.ReadAll(req.Body)` and because of that it will be difficult to create production like behavior.